packages/zero-tag-processor/src/styled.ts

```diff
@@ -0,0 +1,452 @@
+  }[] = [];
+
+  overrides: Record<string, string> = {};
+
+  counterMap: Map<string, number> = new Map();
+
+  baseClasses: string[] = [];
+
+  collectedStyles: [string, string, ExpressionValue | null][] = [];
+
+  collectedVariables: [string, ParseResult<Expression>, boolean][] = [];
+
+  collectedOverrides: [string, string][] = [];
+
+  collectedVariants: VariantDataTransformed[] = [];
+
+  originalLocation: SourceLocation | null = null;
+
+  defaultProps: Record<string, string | number | boolean | unknown> = {};
+
+  constructor(params: Params, ...args: TailProcessorParams) {
+    super(params, ...args);
+    validateParams(params, ['callee', '*', '...'], BaseProcessor.SKIP);
+    validateParams(params, ['callee', ['call', 'member'], ['call', 'template']], 'Invalid params');
+    // console.log(params);
+    const [call, memberOrCall, styleCall] = params;
+    const [callType, componentArg, componentMetaArg] = memberOrCall;
+    const [, ...styleArgs] = styleCall;
+    this.componentMetaArg =
+      componentMetaArg && componentMetaArg.kind === ValueType.LAZY ? componentMetaArg : undefined;
+    this.styleArgs = styleArgs as ExpressionValue[];
+
+    if (callType === 'member') {
+      this.component = componentArg;
+    } else {
+      switch (componentArg.kind) {
+        case ValueType.CONST:
+          this.component = typeof componentArg.value === 'string' ? componentArg.value : undefined;
+          break;
+        case ValueType.LAZY:
+          this.component = {
+            node: componentArg.ex,
+            source: componentArg.source,
+          };
+          this.dependencies.push(componentArg);
+          break;
+        default:
+          this.component = 'FunctionalComponent';
+          break;
+      }
+      if (componentMetaArg && componentMetaArg.kind !== ValueType.FUNCTION) {
+        this.dependencies.push(componentMetaArg);
+      }
+    }
+    if (!this.component) {
+      throw new Error('Invalid usage of `styled` tag');
+    }
+    styleArgs.forEach((item) => {
+      if (!Array.isArray(item)) {
+        if ('kind' in item) {
+          this.dependencies.push(item);
+        }
+      }
+    });
+    if (call[0] === 'callee') {
+      this.originalLocation = call[1].loc ?? null;
+    }
+  }
+
+  getClassName(key?: string) {
+    const mapKey = key ?? 'base';
+    if (!this.counterMap.has(mapKey)) {
+      this.counterMap.set(mapKey, 0);
+    }
+    const currentCount = this.counterMap.get(mapKey) as number;
+    this.counterMap.set(mapKey, currentCount + 1);
+
+    return `${this.className}${mapKey === 'base' ? '' : `-${mapKey}`}${
+      currentCount > 0 ? `-${currentCount}` : ''
+    }`;
+  }
+
+  doEvaltimeReplacement(): void {
+    this.replacer(this.value, false);
+  }
+
+  /**
+   * Order of processing styles -
+   * 1. CSS directly declared in styled call
+   * 2. CSS declared in theme object's styledOverrides
+   * 3. Variants declared in styled call
+   * 3. Variants declared in theme object
+   */
```

Comment:
We will need to review this order. For e.g. now, whatever is defined in the theme would override the styles defined in the component (including "variants" styles). What I mean by this is:

```
const theme = createTheme({
  components: {
    MuiButton: { 
      styleOverrides: { // or variants with props {}
        root: { 
           background: '#00FF00',
        }
      }
    }
  }
});
```

Will override the button's background regardless of what color is used. We can come back to this later tough, I am just mentioning it for awareness.

Reply:
One more question, how do we make sure that the component that is wrapped has it's styles defined previously? In cases like:

```
styled(Slider)
```

Reply:
Variant styles come after `styleOverrides`. So it'll only override base css definitions and then variants will override `styleOverrides`.

Reply:
> One more question...

It's made sure by virtue of the bundler. If a component `A` is imported from `A.ts` like - 

`A.tsx`
```tsx
export const A = styled()...
```

and imported in `B.tsx` 

```tsx
import {} from './A';

const B = styled(A)(...)
```
CSS class ordering is in the order of the import. So first `styled` call in A gets processed and then in B

```css
.class-A {}
.class-B {}
```

Reply:
> Even if you define variants in base css, they get added after styleOverrides css.

This makes sense to me, however is different than the current behavior, which may cause some issues. But again, let's revisit this later and focus on the other aspects.